traffic_ops/traffic_ops_golang/invalidationjobs/invalidationjobs.go

```diff
@@ -0,0 +1,928 @@
+		return
+	}
+
+	input := tc.InvalidationJob{}
+	cType := r.Header.Get(http.CanonicalHeaderKey("content-type"))
+	if cType == tc.ApplicationJson || cType == "" {
+		decoder := json.NewDecoder(r.Body)
+		if err := decoder.Decode(&input); err != nil {
+			userErr = fmt.Errorf("Unable to parse input: %v", err)
+			sysErr = fmt.Errorf("parsing input to PUT jobs?id=%s: %v", inf.Params["id"], err)
+			errCode = http.StatusBadRequest
+			api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+			return
+		}
+	} else {
+		w.Header().Set(http.CanonicalHeaderKey("accept"), strings.Join([]string{tc.ApplicationJson, "application/x-www-form-urlencoded", "multipart/form-data"}, ";"))
+		err := fmt.Errorf("unsupported content-type: %s", cType)
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusUnsupportedMediaType, err, err)
+		return
+	}
+
+	if err := input.Validate(); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, err, nil)
+		return
+	}
+
+	if job.StartTime.Before(time.Now()) {
+		userErr = errors.New("Cannot modify a job that has already started!")
+		errCode = http.StatusMethodNotAllowed
+		w.Header().Set(http.CanonicalHeaderKey("allow"), "GET,HEAD,DELETE")
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+	if *job.DeliveryService != *input.DeliveryService {
+		userErr = errors.New("Cannot change 'deliveryService' of existing invalidation job!")
+		errCode = http.StatusConflict
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+	if *job.CreatedBy != *input.CreatedBy {
+		userErr = errors.New("Cannot change 'createdBy' of existing invalidation jobs!")
+		errCode = http.StatusConflict
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+	if *job.ID != *input.ID {
+		userErr = errors.New("Cannot change an invalidation job 'id'!")
+		errCode = http.StatusConflict
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, nil)
+		return
+	}
+
+
+	row = inf.Tx.Tx.QueryRow(updateQuery,
+		input.AssetURL,
+		input.Keyword,
+		input.Parameters,
+		input.StartTime.Time,
+		*job.ID)
+	err = row.Scan(&job.AssetURL,
+		&job.CreatedBy,
+		&job.DeliveryService,
+		&job.ID,
+		&job.Keyword,
+		&job.Parameters,
+		&job.StartTime)
+	if err != nil {
+		sysErr = fmt.Errorf("Updating a job: %v", err)
+		errCode = http.StatusInternalServerError
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, nil, sysErr)
+		return
+	}
+
+	if err = setRevalFlags(*job.DeliveryService, inf.Tx.Tx); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("Setting reval flags: %v", err))
+		return
+	}
+
+	response := apiResponse{
+		[]tc.Alert{
+			tc.Alert{
+				"Content invalidation job updated",
+				tc.SuccessLevel.String(),
+			},
+		},
+		job,
+	}
+
+	resp, err := json.Marshal(response)
+	if err != nil {
+		sysErr = fmt.Errorf("encoding response: %v", err)
+		errCode = http.StatusInternalServerError
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, nil, sysErr)
```

Comment:
return here?

Reply:
Actually, I don't return here so that the changelog is created whether or not an encoding error occurs. If this error occurs, the job has already been created, so it's too late to bail.

Reply:
we roll back though https://github.com/apache/trafficcontrol/blob/master/traffic_ops/traffic_ops_golang/api/api.go#L111?

Reply:
yeah that ^ :)